File: connect/src/main/java/org/transdroid/connect/clients/ClientDelegate.java
package org.transdroid.connect.clients;

import org.transdroid.connect.model.Torrent;

import java.io.InputStream;

import io.reactivex.Flowable;

/**
 * Wraps an actual client implementation by calling through the appropriate method only if it is supported. This allows the final
 * {@link ClientSpec} API to expose all methods without forcing the individual implementations to implement unsupported featured with a no-op.
 */
final class ClientDelegate implements ClientSpec {

	private final Client client;
	private final Object actual;

	ClientDelegate(Client client, Object actual) {
		this.client = client;
		this.actual = actual;
	}

	@Override
	public Flowable<Torrent> torrents() {
		if (client.supports(Feature.LISTING))
			return ((Feature.Listing) actual).torrents();
		throw new UnsupportedFeatureException(client, Feature.LISTING);
	}

	@Override
	public Flowable<String> clientVersion() {
		if (client.supports(Feature.VERSION))
			return ((Feature.Version) actual).clientVersion();
		throw new UnsupportedFeatureException(client, Feature.VERSION);
	}

	@Override
	public Flowable<Torrent> start(Torrent torrent) {
		if (client.supports(Feature.STARTING_STOPPING))
			return ((Feature.StartingStopping) actual).start(torrent);
		throw new UnsupportedFeatureException(client, Feature.STARTING_STOPPING);
	}

	@Override
	public Flowable<Torrent> stop(Torrent torrent) {
		if (client.supports(Feature.STARTING_STOPPING))
			return ((Feature.StartingStopping) actual).stop(torrent);
		throw new UnsupportedFeatureException(client, Feature.STARTING_STOPPING);
	}

	@Override
	public Flowable<Torrent> forceStart(Torrent torrent) {
		if (client.supports(Feature.FORCE_STARTING))
			return ((Feature.ForceStarting) actual).forceStart(torrent);
		throw new UnsupportedFeatureException(client, Feature.FORCE_STARTING);
	}

	@Override
	public Flowable<Void> addByFile(InputStream file) {
		if (client.supports(Feature.ADD_BY_FILE))
			return ((Feature.AddByFile) actual).addByFile(file);
		throw new UnsupportedFeatureException(client, Feature.ADD_BY_FILE);
	}

	@Override
	public Flowable<Void> addByUrl(String url) {
		if (client.supports(Feature.ADD_BY_URL))
			return ((Feature.AddByUrl) actual).addByUrl(url);
		throw new UnsupportedFeatureException(client, Feature.ADD_BY_URL);
	}

	@Override
	public Flowable<Void> addByMagnet(String magnet) {
		if (client.supports(Feature.ADD_BY_MAGNET))
			return ((Feature.AddByMagnet) actual).addByMagnet(magnet);
		throw new UnsupportedFeatureException(client, Feature.ADD_BY_MAGNET);
	}

}

File: connect/src/main/java/org/transdroid/connect/clients/ClientDelegate.kt
package org.transdroid.connect.clients

import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import org.transdroid.connect.model.Torrent
import java.io.InputStream

/**
 * Wraps an actual client implementation by calling through the appropriate method only if it is supported. This allows the final
 * [ClientSpec] API to expose all methods without forcing the individual implementations to implement unsupported featured with a no-op.
 */
internal class ClientDelegate(private val client: Client, private val actual: Any) : ClientSpec {

    override fun torrents(): Flowable<Torrent> {
        if (client.supports(Feature.LISTING))
            return (actual as Feature.Listing).torrents()
        throw UnsupportedFeatureException(client, Feature.LISTING)
    }

    override fun clientVersion(): Single<String> {
        if (client.supports(Feature.VERSION))
            return (actual as Feature.Version).clientVersion()
        throw UnsupportedFeatureException(client, Feature.VERSION)
    }

    override fun start(torrent: Torrent): Single<Torrent> {
        if (client.supports(Feature.STARTING_STOPPING))
            return (actual as Feature.StartingStopping).start(torrent)
        throw UnsupportedFeatureException(client, Feature.STARTING_STOPPING)
    }

    override fun stop(torrent: Torrent): Single<Torrent> {
        if (client.supports(Feature.STARTING_STOPPING))
            return (actual as Feature.StartingStopping).stop(torrent)
        throw UnsupportedFeatureException(client, Feature.STARTING_STOPPING)
    }

    override fun forceStart(torrent: Torrent): Single<Torrent> {
        if (client.supports(Feature.FORCE_STARTING))
            return (actual as Feature.ForceStarting).forceStart(torrent)
        throw UnsupportedFeatureException(client, Feature.FORCE_STARTING)
    }

    override fun addByFile(file: InputStream): Completable {
        if (client.supports(Feature.ADD_BY_FILE))
            return (actual as Feature.AddByFile).addByFile(file)
        throw UnsupportedFeatureException(client, Feature.ADD_BY_FILE)
    }

    override fun addByUrl(url: String): Completable {
        if (client.supports(Feature.ADD_BY_URL))
            return (actual as Feature.AddByUrl).addByUrl(url)
        throw UnsupportedFeatureException(client, Feature.ADD_BY_URL)
    }

    override fun addByMagnet(magnet: String): Completable {
        if (client.supports(Feature.ADD_BY_MAGNET))
            return (actual as Feature.AddByMagnet).addByMagnet(magnet)
        throw UnsupportedFeatureException(client, Feature.ADD_BY_MAGNET)
    }

}
